src/coins.cpp

```diff
@@ -152,36 +165,35 @@ bool CCoinsViewCache::BatchWrite(CCoinsMap &mapCoins, const uint256 &hashBlockIn
         }
         CCoinsMap::iterator itUs = cacheCoins.find(it->first);
         if (itUs == cacheCoins.end()) {
-            // The parent cache does not have an entry, while the child does
-            // We can ignore it if it's both FRESH and pruned in the child
+            // The parent cache does not have an entry, while the child cache does.
+            // We can ignore it if it's both spent and FRESH in the child
             if (!(it->second.flags & CCoinsCacheEntry::FRESH && it->second.coin.IsSpent())) {
-                // Otherwise we will need to create it in the parent
-                // and move the data up and mark it as dirty
+                // Create the coin in the parent cache, move the data up
+                // and mark it as dirty.
                 CCoinsCacheEntry& entry = cacheCoins[it->first];
                 entry.coin = std::move(it->second.coin);
                 cachedCoinsUsage += entry.coin.DynamicMemoryUsage();
                 entry.flags = CCoinsCacheEntry::DIRTY;
-                // We can mark it FRESH in the parent if it was FRESH in the child
-                // Otherwise it might have just been flushed from the parent's cache
-                // and already exist in the grandparent
+                // A coin can only be FRESH in the child cache if it doesn't exist
+                // in any of the ancestor caches. We can therefore mark it FRESH
```

Comment:
In commit "[docs] Improve commenting in coins.cpp|h" (eeeb97b6da16fed5701ff04d3b9a622eb759fa5a)

"A coin can only be FRESH in the child cache if it doesn't exist in any of the ancestor caches" is too broad a statement, not generally true. You can layer caches arbitrarily and a fresh flag in a child doesn't tell you about the state of the coin in every ancestor cache. It only directly tells you the state of the coin in the immediate parent cache. And in this case, we indirectly know the state of the grandparent coin due to this happening under the `itUs == cacheCoins.end()` branch.

Would write this comment more like: "We can mark parent cache entry FRESH here if child cache entry was FRESH. This works because parent cache entry not previously existing and child cache entry being FRESH indicates grandparent coin is spent or doesn't exist."

Reply:
> In this case, we only indirectly know the state of the grandparent coin due to this happening under the itUs == cacheCoins.end() branch.

I don't understand this statement. `itUs == cacheCoins.end()` simply tells us that the coin doesn't exist in the parent cache's `CCoinsMap` structure. It doesn't tell us anything about the grandparent cache.

Reply:
> > In this case, we only indirectly know the state of the grandparent coin due to this happening under the itUs == cacheCoins.end() branch.

If entry has fresh flag and there is no parent entry (`itUs == cacheCoins.end()`) then the coin must be spent in the grandparent and therefore it's safe to re-apply the fresh flag

> It doesn't tell us anything about the grandparent cache.

Of course, the grandparent might not even be a cache view. But whatever kind of view it is the coin must be spent there.

Reply:
> If entry has fresh flag and there is no parent entry (itUs == cacheCoins.end()) then the coin must be spent in the grandparent

I still don't understand this. Can you explain the logic?

(I still believe that no ancestor cache can contain the coin due to the way we build the cache hierarchy, but I don't understand how you reason about just the grandparent cache)

Reply:
I don't see why we're going back and forth on one sentence in a code review comment I made explaining why I thought your code comment was making an unstated assumption. If you don't like my suggestion (which does not mention itUs), feel free to ignore!

My logic is just that in order for a cache entry to wind up with the fresh flag, the coin in its parent view (whether it's a cache or not, whether or not it has entries or is flushed if it is a cache) must be spent. If the parent view is a cache, and doesn't have an entry, the coin must be spent in the grandparent view (again whether or not it's a cache and regardless of whether entries are created).

I am using "spent" here as shorthand for "spent or never existed" in case that what may be confusing

Reply:
Maybe: "A coin can only be FRESH in the child cache if it doesn't exist (e.g. was spent or never existed) in any of the ancestor caches".

Reply:
or: "A coin can only be FRESH in the child cache if it doesn't exist (e.g. was spent or never existed) in the parent view, nor in any of the ancestor caches due to the way we build the cache hierarchy."

Reply:
> I don't see why we're going back and forth on one sentence in a code review comment I made explaining why I thought your code comment was making an unstated assumption. 

Because I'm trying to understand your logic and if there's a gap in my understanding. And if not, whether the comment is needlessly confusing.

> in order for a cache entry to wind up with the fresh flag, the coin in its parent view [...] must be spent. If the parent view is a cache, and doesn't have an entry, the coin must be spent in the grandparent view 

But isn't this transitive? Can I not equally say "If the grandparent view is a cache, and doesn't have an entry, the coin must be spent in the great-grandparent view" and so on?